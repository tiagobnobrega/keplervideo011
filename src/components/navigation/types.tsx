/*
 * Copyright (c) 2024 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */

import { DrawerScreenProps } from '@amzn/react-navigation__drawer/lib/typescript/src/types';
import { StackNavigationProp } from '@amzn/react-navigation__stack';
import { StackScreenProps } from '@amzn/react-navigation__stack/lib/typescript/src/types';
import { ImageSourcePropType, TouchableOpacity } from 'react-native';
import { TitleData } from '../../types/TitleData';

export enum Screens {
  DEFAULT_SCREEN = 'Home',
  APP_DRAWER = 'AppDrawer',
  HOME_SCREEN = 'Home',
  DETAILS_SCREEN = 'Details',
  PLAYER_SCREEN = 'Player',
  SETTINGS_SCREEN = 'Settings',
  SEARCH_SCREEN = 'Search',
  SEARCH_RESULTS_SCREEN = 'SearchResultsScreen',
  FEEDBACK_SCREEN = 'FeedBackScreen',
}

export enum DrawerType {
  PERMANENT = 'permanent',
  FRONT = 'front',
  SLIDE = 'slide',
  BACK = 'back',
}

export type AppStackParamList = {
  [Screens.APP_DRAWER]: undefined;
  [Screens.HOME_SCREEN]: undefined;
  [Screens.SEARCH_SCREEN]: undefined;
  [Screens.SETTINGS_SCREEN]: undefined;
  [Screens.FEEDBACK_SCREEN]: undefined;
  [Screens.SEARCH_RESULTS_SCREEN]: {
    searchKeyword: string;
  };
  [Screens.DETAILS_SCREEN]: {
    data: TitleData;
    sendDataOnBack: (id: number | string) => void;
  };
  [Screens.PLAYER_SCREEN]: {
    data: TitleData;
    sendDataOnBack: () => void;
  };
};

export type AppStackScreenProps<T extends keyof AppStackParamList> =
  StackScreenProps<AppStackParamList, T>;

export type HomeScreenNavigationProps = StackNavigationProp<
  AppStackParamList,
  Screens.HOME_SCREEN
>;

export type AppDrawerParamList = {
  [Screens.HOME_SCREEN]: undefined;
  [Screens.SETTINGS_SCREEN]: undefined;
  [Screens.SEARCH_SCREEN]: undefined;
};

export type AppDrawerScreenProps<T extends keyof AppDrawerParamList> =
  DrawerScreenProps<AppDrawerParamList, T>;

declare global {
  namespace ReactNavigation {
    interface RootParamList extends AppStackParamList {}
  }
}

export interface ButtonConfig {
  onPress: () => void;
  image: ImageSourcePropType;
  label: string;
  ref?: React.RefObject<TouchableOpacity>;
  testID: string;
}
