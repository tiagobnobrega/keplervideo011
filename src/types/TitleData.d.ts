import { MediaItem } from '@amzn/keplerblocks-types';

export type MimeType =
  | 'text/plain'
  | 'text/vtt'
  | 'application/x-subtitle' // .srt
  | 'application/x-subtitle-vtt' // .vtt
  | 'application/x-subtitle-sami'
  | 'application/x-subtitle-tmplayer'
  | 'application/x-subtitle-mpl2'
  | 'application/x-subtitle-dks'
  | 'application/x-subtitle-qttext'
  | 'application/x-subtitle-lrc';

export type VideoFormat = 'MPD' | 'HLS' | 'MP4' | 'DASH';

interface DrmScheme {
  name: string;
  licenseUri: string;
  headerTag?: string;
  headerData?: string;
}
interface TextTrack {
  label: string;
  language: string;
  uri: string;
  mimeType: MimeType;
}

export interface TitleData extends MediaItem {
  videoUrl: string;
  categories: string[];
  channelID: string;
  posterUrl: string;
  format: string;
  drmScheme?: DrmScheme;
  textTrack?: TextTrack[];
  uhd: boolean;
  secure: boolean;
  aCodec?: string;
  vCodec?: string;
  rentAmount: string;
}
