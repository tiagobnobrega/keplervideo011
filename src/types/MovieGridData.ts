import { TitleData } from './TitleData';

export type MovieGridData = {
  heading: string;
  testID: string;
  data: () => TitleData[];
};
