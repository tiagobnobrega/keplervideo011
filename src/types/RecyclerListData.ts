import { StyleProp, ViewStyle } from 'react-native';
import { RecyclerListViewProps } from 'recyclerlistview';

export interface CustomRecyclerListViewProps extends RecyclerListViewProps {
  contentContainerStyle?: StyleProp<ViewStyle>;
  initialNumToRender: number | undefined;
}
