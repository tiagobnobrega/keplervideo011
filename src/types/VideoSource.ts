// adapted from GStreamer MIME Types (https://gstreamer.freedesktop.org/documentation/subparse/subparse.html?gi-language=c)
export type MimeType =
  | 'text/plain'
  | 'text/vtt'
  | 'application/x-subtitle' // .srt
  | 'application/x-subtitle-vtt' // .vtt
  | 'application/x-subtitle-sami'
  | 'application/x-subtitle-tmplayer'
  | 'application/x-subtitle-mpl2'
  | 'application/x-subtitle-dks'
  | 'application/x-subtitle-qttext'
  | 'application/x-subtitle-lrc';

export type VideoFormat = 'MPD' | 'HLS' | 'MP4' | 'DASH';

interface DrmScheme {
  name: string;
  licenseUri: string;
  headerTag?: string;
  headerData?: string;
}
interface TextTrack {
  label: string;
  language: string;
  uri: string;
  mimeType: MimeType;
}

export interface VideoSource {
  title: string;
  uri: string;
  format: VideoFormat;
  drmScheme?: DrmScheme;
  textTrack?: TextTrack[];
  uhd: boolean;
  secure: boolean;
  aCodec?: string;
  vCodec?: string;
}
