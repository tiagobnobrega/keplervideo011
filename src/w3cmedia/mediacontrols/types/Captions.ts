import { VideoPlayer } from '@amzn/react-native-w3cmedia';

export interface CaptionButtonProps {
  onPress: () => void;
  video: VideoPlayer;
  captionMenuVisibility: boolean;
  testID: string;
  playerControlType?: string;
  isCaptionButtonFocused: boolean | null;
}

export interface CaptionOptionProps {
  video: VideoPlayer;
  enableCaption: (captionId: string) => void;
  selectedCaptionId: string | null;
}

export interface CaptionMenuItemProps {
  text: string;
  index?: string;
  selected?: boolean;
  onPress: () => void;
  testID?: string;
}

export interface CaptionMenuProps {
  captionMenuVisibility: boolean;
  video: VideoPlayer;
  setSelectedCaption: (id: string) => void;
}
