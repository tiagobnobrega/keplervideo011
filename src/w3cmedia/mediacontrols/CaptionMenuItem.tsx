import React from 'react';
import { StyleSheet, Text } from 'react-native';
import FocusableElement from '../../components/FocusableElement';
import { COLORS } from '../../styles/Colors';
import { scaleUxToDp } from '../../utils/pixelUtils';
import { CaptionMenuItemProps } from './types/Captions';

const CaptionMenuItem = React.memo(
  ({ text, selected, onPress, testID }: CaptionMenuItemProps) => {
    return (
      <FocusableElement
        onPress={onPress}
        style={selected && styles.selectedMenuItem}
        onFocusOverrideStyle={styles.focusedMenuItem}
        testID={testID}>
        <Text style={styles.menuItemText} testID={'video-player-caption-menu'}>
          {text}
        </Text>
      </FocusableElement>
    );
  },
);
export default CaptionMenuItem;

const styles = StyleSheet.create({
  menuItemText: {
    color: COLORS.WHITE,
    fontWeight: '500',
    fontSize: scaleUxToDp(30),
    marginVertical: scaleUxToDp(4),
    padding: scaleUxToDp(20),
  },
  focusedMenuItem: { backgroundColor: COLORS.GRAY + 'E6' },
  selectedMenuItem: { backgroundColor: COLORS.GRAY + '66' },
});
