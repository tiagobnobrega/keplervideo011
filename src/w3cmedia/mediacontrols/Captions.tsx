import { TVFocusGuideView } from '@amzn/react-native-kepler';
import React, { useCallback, useState } from 'react';
import { StyleSheet, View } from 'react-native';
import { CAPTION_DISABLE_ID, CAPTION_DISABLE_TEXT } from '../../constants';
import { COLORS } from '../../styles/Colors';
import { scaleUxToDp } from '../../utils/pixelUtils';
import CaptionMenuItem from './CaptionMenuItem';
import CaptionOptions from './CaptionOptions';
import { CaptionMenuProps } from './types/Captions';

export const CaptionMenu = ({
  captionMenuVisibility,
  video,
  setSelectedCaption,
}: CaptionMenuProps) => {
  const [captionID, setCaptionID] = useState<string>(CAPTION_DISABLE_ID);

  const removeCurrentCaption = useCallback(() => {
    if (captionID !== CAPTION_DISABLE_ID) {
      console.log('Disabling Captions');
      if (video && video.textTracks) {
        video.textTracks.getTrackById(captionID)!.mode = 'hidden';
      }
    }
    setCaptionID(CAPTION_DISABLE_ID);
  }, [captionID, video]);

  const enableCaption = useCallback(
    (id: string) => {
      if (id === captionID) {
        return;
      }
      removeCurrentCaption();
      if (video && video.textTracks) {
        video.textTracks.getTrackById(id)!.mode = 'showing';
        setCaptionID(id);
        setSelectedCaption(id);
      }
      setCaptionID(id);
      setSelectedCaption(id);
    },
    [captionID, removeCurrentCaption, video, setSelectedCaption],
  );

  const turnOffCaptions = useCallback(() => {
    setSelectedCaption(CAPTION_DISABLE_ID);
    removeCurrentCaption();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <TVFocusGuideView trapFocusLeft trapFocusRight trapFocusUp>
      {captionMenuVisibility ? (
        <View style={styles.menu}>
          <CaptionOptions
            video={video}
            enableCaption={enableCaption}
            selectedCaptionId={captionID}
          />
          <CaptionMenuItem
            key={CAPTION_DISABLE_TEXT}
            text={CAPTION_DISABLE_TEXT}
            onPress={turnOffCaptions}
            selected={captionID === CAPTION_DISABLE_ID}
            testID={'caption-menu-focus-element'}
          />
        </View>
      ) : null}
    </TVFocusGuideView>
  );
};

const styles = StyleSheet.create({
  menu: {
    backgroundColor: COLORS.BLACK + 'E6',
    width: scaleUxToDp(250),
    justifyContent: 'center',
    zIndex: 30,
    borderRadius: scaleUxToDp(15),
    overflow: 'hidden',
  },
  captionSelected: {
    backgroundColor: COLORS.DARK_GRAY + 'D9',
    borderRadius: scaleUxToDp(40),
  },
});
