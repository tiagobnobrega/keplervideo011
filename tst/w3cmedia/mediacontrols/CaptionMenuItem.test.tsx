/*
 * Copyright (c) 2024 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */
import { describe } from '@jest/globals';
import '@testing-library/jest-native/extend-expect';
import { render } from '@testing-library/react-native';
import React from 'react';
import CaptionMenuItem from '../../../src/w3cmedia/mediacontrols/CaptionMenuItem';
import { CaptionMenuItemProps } from '../../../src/w3cmedia/mediacontrols/types/Captions';

const onPressMock = jest.fn();
const renderCaptionButton = (props?: Partial<CaptionMenuItemProps>) => {
  return (
    <CaptionMenuItem text={'Some Text'} onPress={onPressMock} {...props} />
  );
};
describe('CaptionMenuItem tests', () => {
  it('component renders correctly', () => {
    const component = render(renderCaptionButton());
    expect(component).toMatchSnapshot();
  });
  it('component renders correctly with select', () => {
    const component = render(renderCaptionButton({ selected: true }));
    expect(component).toMatchSnapshot();
  });
});
