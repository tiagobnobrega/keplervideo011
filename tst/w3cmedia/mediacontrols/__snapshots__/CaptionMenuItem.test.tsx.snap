// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CaptionMenuItem tests component renders correctly 1`] = `
<View
  accessibilityState={
    Object {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    Object {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#FFFFFF",
        "fontSize": 30,
        "fontWeight": "500",
        "marginVertical": 4,
        "padding": 20,
      }
    }
    testID="video-player-caption-menu"
  >
    Some Text
  </Text>
</View>
`;

exports[`CaptionMenuItem tests component renders correctly with select 1`] = `
<View
  accessibilityState={
    Object {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    Object {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#80808066",
      "opacity": 1,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#FFFFFF",
        "fontSize": 30,
        "fontWeight": "500",
        "marginVertical": 4,
        "padding": 20,
      }
    }
    testID="video-player-caption-menu"
  >
    Some Text
  </Text>
</View>
`;
